### PR TITLE
Adding updateAll method to SPI

### PR DESCRIPTION
**Pull request by sanjanasandeep**
Adding this method to update to DELETED status using a single SQL query during async deletion 

JIRA: https://cdap.atlassian.net/browse/CDAP-20306

### CODE REVIEW

File: cdap-data-fabric/src/main/java/io/cdap/cdap/spi/data/common/MetricStructuredTable.java [hunk 1]
```diff
@@ -303,6 +303,24 @@ public void deleteAll(Range keyRange) throws InvalidFieldException, IOException
     }
   }
```
Comment (sanjanasandeep) [inline cdap-data-fabric/src/main/java/io/cdap/cdap/spi/data/common/MetricStructuredTable.java:305]: Adding unit test in next rev

File: cdap-storage-spi/src/main/java/io/cdap/cdap/spi/data/StructuredTable.java [hunk 2]
```diff
@@ -273,6 +273,17 @@ void increment(Collection<Field<?>> keys, String column, long amount)
    */
   void deleteAll(Range keyRange) throws InvalidFieldException, IOException;
 
+  /**
+   * Updates the specific fields in a range of rows from the table
+   *
+   * @param keyRange key range of the rows to delete
```
Comment (de-lan) [inline cdap-storage-spi/src/main/java/io/cdap/cdap/spi/data/StructuredTable.java]: `to delete` --> `to update`

File: cdap-storage-spi/src/main/java/io/cdap/cdap/spi/data/StructuredTable.java [hunk 3]
```diff
@@ -273,6 +273,17 @@ void increment(Collection<Field<?>> keys, String column, long amount)
...
+   * @param fields the fields to write to
+   * @throws InvalidFieldException if any of the keys are not part of table schema,
+   *                               or their types do not match the schema
+   * @throws IOException if there is an error reading or deleting from the table
```
Comment (de-lan) [inline cdap-storage-spi/src/main/java/io/cdap/cdap/spi/data/StructuredTable.java]: `reading or deleting from the table` --> `updating from the table`

File: cdap-storage-ext-spanner/src/main/java/io/cdap/cdap/storage/spanner/SpannerStructuredTable.java [hunk 4]
```diff
@@ -367,6 +368,28 @@ public void deleteAll(Range range) throws InvalidFieldException {
     transactionContext.executeUpdate(builder.build());
   }
 
+  @Override
+  public void updateAll(Range keyRange, Collection<Field<?>> fields) throws InvalidFieldException, IOException {
+    fieldValidator.validatePrimaryKeys(keyRange.getBegin(), true);
```
Comment (de-lan) [inline cdap-storage-ext-spanner/src/main/java/io/cdap/cdap/storage/spanner/SpannerStructuredTable.java:373]: I agree the `keyRange` must be primary keys in order. Let's also specify it in the interface javaDoc 

Also pls check the passed in `fields` is a valid field (`fieldValidator::validateField`) but not a primary key.

File: cdap-storage-ext-spanner/src/main/java/io/cdap/cdap/storage/spanner/SpannerStructuredTable.java [hunk 5]
```diff
@@ -367,6 +368,28 @@ public void deleteAll(Range range) throws InvalidFieldException {
...
+    fieldValidator.validatePrimaryKeys(keyRange.getEnd(), true);
+
+    Map<String, Value> parameters = new HashMap<>();
```
Comment (de-lan) [inline cdap-storage-ext-spanner/src/main/java/io/cdap/cdap/storage/spanner/SpannerStructuredTable.java:378]: Did we bind these `parameters` for the WHERE clause? Something like `parameters.forEach((name, value) -> builder.bind(name).to(value));`

File: cdap-data-fabric/src/main/java/io/cdap/cdap/spi/data/sql/PostgreSqlStructuredTable.java [hunk 6]
```diff
@@ -542,6 +542,13 @@ public void deleteAll(Range keyRange) throws InvalidFieldException, IOException
     }
   }
 
+  @Override
+  public void updateAll(Range keyRange, Collection<Field<?>> fields) throws InvalidFieldException, IOException {
+    LOG.trace("Table {}: Update fields {} in range {}", tableSchema.getTableId(), fields, keyRange);
+    fieldValidator.validateScanRange(keyRange);
```
Comment (de-lan) [inline cdap-data-fabric/src/main/java/io/cdap/cdap/spi/data/sql/PostgreSqlStructuredTable.java]: Same comment from the Spanner implementation, pls specify the `keyRange` needs to be primary keys in order and make sure that the passed fields are all valid non-key fields

File: cdap-data-fabric/src/main/java/io/cdap/cdap/spi/data/sql/PostgreSqlStructuredTable.java [hunk 7]
```diff
@@ -590,6 +597,24 @@ private void upsertInternal(Collection<Field<?>> fields) throws IOException {
...
+  private void updateRangeInternal(Range keyRange, Collection<Field<?>> fields) throws IOException {
+    String sqlQuery = getUpdateRangeSqlQuery(keyRange, fields);
+    try (PreparedStatement statement = connection.prepareStatement(sqlQuery)) {
+      int index = 1;
+      // set field values to update
+      for (Field<?> field : fields) {
+        setField(statement, field, index);
+        index++;
+      }
+      setStatementFieldByRange(keyRange, statement);
```
Comment (de-lan) [inline cdap-data-fabric/src/main/java/io/cdap/cdap/spi/data/sql/PostgreSqlStructuredTable.java]: I think it should be `setStatementFieldByRange(keyRange, statement, index);` Since we've set values for the fields to update, we should skip to the range part

File: cdap-data-fabric/src/main/java/io/cdap/cdap/spi/data/nosql/NoSqlStructuredTable.java [hunk 8]
```diff
@@ -431,6 +431,21 @@ public void deleteAll(Range keyRange) throws InvalidFieldException, IOException
     }
   }
 
+  @Override
+  public void updateAll(Range keyRange, Collection<Field<?>> fields) throws InvalidFieldException, IOException {
+    LOG.trace("Table {}: UpdateAll with range {}", schema.getTableId(), keyRange);
+    try (CloseableIterator<StructuredRow> iterator = getFilterByRangeIterator(keyRange)) {
```
Comment (de-lan) [inline cdap-data-fabric/src/main/java/io/cdap/cdap/spi/data/nosql/NoSqlStructuredTable.java:443]: Same, make sure the keyRange is primary keys only and fields are non-key valid fields

File: cdap-data-fabric/src/main/java/io/cdap/cdap/spi/data/nosql/NoSqlStructuredTable.java [hunk 9]
```diff
@@ -431,6 +431,21 @@ public void deleteAll(Range keyRange) throws InvalidFieldException, IOException
...
+      while (iterator.hasNext()) {
+        Collection<Field<?>> primaryKeys = iterator.next().getPrimaryKeys();
+        Put put = updateFieldsToBytes(primaryKeys);
```
Comment (de-lan) [inline cdap-data-fabric/src/main/java/io/cdap/cdap/spi/data/nosql/NoSqlStructuredTable.java]: Is the fields to update even used?

Comment (sanjanasandeep) [reply cdap-data-fabric/src/main/java/io/cdap/cdap/spi/data/nosql/NoSqlStructuredTable.java]: yeah need to update L439. I feel like there are lot of things missed without testing. Let me add the test first and then we can automatically catch these errors.

File: cdap-storage-ext-spanner/src/main/java/io/cdap/cdap/storage/spanner/SpannerStructuredTable.java (hunk 5 above)
Comment (sanjanasandeep) [reply cdap-storage-ext-spanner/src/main/java/io/cdap/cdap/storage/spanner/SpannerStructuredTable.java:378]: this makes more sense after debugging. Good catch, thanks!

File: cdap-storage-ext-spanner/src/test/java/io/cdap/cdap/storage/spanner/SpannerStructuredTableTest.java [hunk 10]
```diff
@@ -50,10 +50,10 @@ public class SpannerStructuredTableTest extends StructuredTableTest {
 
   @BeforeClass
   public static void createSpannerStorageProvider() throws Exception {
-    String project = System.getProperty("gcp.project");
-    String instance = System.getProperty("gcp.spanner.instance");
-    String database = System.getProperty("gcp.spanner.database");
-    String credentialsPath = System.getProperty("gcp.credentials.path");
+    String project = "skilled-mark-331618";
```
Comment (de-lan) [inline cdap-storage-ext-spanner/src/test/java/io/cdap/cdap/storage/spanner/SpannerStructuredTableTest.java]: pls revert these changes, currently SpannerTable tests are running in real env so they're skipped in unit tests

File: cdap-data-fabric/src/main/java/io/cdap/cdap/spi/data/nosql/NoSqlStructuredTable.java [hunk 11]
```diff
@@ -431,6 +432,29 @@ public void deleteAll(Range keyRange) throws InvalidFieldException, IOException
     }
   }
 
+  @Override
+  public void updateAll(Range keyRange, Collection<Field<?>> fields) throws InvalidFieldException {
+    LOG.trace("Table {}: Update fields {} in range {}", schema.getTableId(), fields, keyRange);
+    // validate that the range is strictly a primary key prefix
+    fieldValidator.validatePrimaryKeys(keyRange.getBegin(), true);
+    fieldValidator.validatePrimaryKeys(keyRange.getEnd(), true);
+    fieldValidator.validateScanRange(keyRange);
```
Comment (de-lan) [inline cdap-data-fabric/src/main/java/io/cdap/cdap/spi/data/nosql/NoSqlStructuredTable.java]: This check is not necessary, since above `validatePrimaryKeys` already made sure the range are all keys and prefixed

File: cdap-storage-spi/src/main/java/io/cdap/cdap/spi/data/table/field/FieldValidator.java [hunk 12]
```diff
@@ -148,4 +148,21 @@ public void validatePartialPrimaryKeys(Collection<Field<?>> keys) throws Invalid
...
+   * @param fields the fields to validate
+   * @throws InvalidFieldException if any of the given fields are in the primary key.
+   */
+  public void validateNotPrimaryKeys(Collection<Field<?>> fields) throws InvalidFieldException {
+    Set<String> primaryKeys = new HashSet<>(tableSchema.getPrimaryKeys());
+    for (Field<?> field : fields) {
+      validateField(field);
+      if (primaryKeys.contains(field.getName())) {
+        throw new InvalidFieldException(
+          tableSchema.getTableId(), String.format("Given field % is a primary key which shouldn't be updated", field));
```
Comment (de-lan) [inline cdap-storage-spi/src/main/java/io/cdap/cdap/spi/data/table/field/FieldValidator.java]: nit: `%` --> `%s`

File: cdap-data-fabric/src/main/java/io/cdap/cdap/spi/data/sql/PostgreSqlStructuredTable.java [hunk 13]
```diff
@@ -542,6 +542,18 @@ public void deleteAll(Range keyRange) throws InvalidFieldException, IOException
     }
   }
 
+  @Override
+  public void updateAll(Range keyRange, Collection<Field<?>> fields) throws InvalidFieldException, IOException {
+    LOG.trace("Table {}: Update fields {} in range {}", tableSchema.getTableId(), fields, keyRange);
+    // validate that the range is strictly a primary key prefix
+    fieldValidator.validatePrimaryKeys(keyRange.getBegin(), true);
+    fieldValidator.validatePrimaryKeys(keyRange.getEnd(), true);
+    fieldValidator.validateScanRange(keyRange);
```
Comment (de-lan) [inline cdap-data-fabric/src/main/java/io/cdap/cdap/spi/data/sql/PostgreSqlStructuredTable.java]: Same, this check is not necessary

File: cdap-data-fabric/src/main/java/io/cdap/cdap/spi/data/sql/PostgreSqlStructuredTable.java [hunk 14]
```diff
@@ -590,6 +602,24 @@ private void upsertInternal(Collection<Field<?>> fields) throws IOException {
...
+      // set field values to update
+      for (Field<?> field : fields) {
+        setField(statement, field, index);
+        index++;
+      }
+      setStatementFieldByRange(keyRange, statement, index);
+      LOG.trace("SQL statement: {}", statement);
+      statement.executeUpdate();
+    } catch (SQLException e) {
+      throw new IOException(String.format("Failed to write to table %s with fields %s",
```
Comment (de-lan) [inline cdap-data-fabric/src/main/java/io/cdap/cdap/spi/data/sql/PostgreSqlStructuredTable.java]: nit: could make the error message more precise: `String.format("Failed to update table %s with range %s and fields %s",tableSchema.getTableId().getName(), keyRange, fields)`

File: cdap-storage-ext-spanner/src/main/java/io/cdap/cdap/storage/spanner/SpannerStructuredTable.java [hunk 15]
```diff
@@ -367,6 +367,33 @@ public void deleteAll(Range range) throws InvalidFieldException {
     transactionContext.executeUpdate(builder.build());
   }
 
+  @Override
+  public void updateAll(Range keyRange, Collection<Field<?>> fields) throws InvalidFieldException {
+    // validate that the range is strictly a primary key prefix
+    fieldValidator.validatePrimaryKeys(keyRange.getBegin(), true);
+    fieldValidator.validatePrimaryKeys(keyRange.getEnd(), true);
+    fieldValidator.validateScanRange(keyRange);
```
Comment (de-lan) [inline cdap-storage-ext-spanner/src/main/java/io/cdap/cdap/storage/spanner/SpannerStructuredTable.java]: Same, no need to `validateScanRange`

File: cdap-storage-spi/src/main/java/io/cdap/cdap/spi/data/StructuredTable.java [hunk 16]
```diff
@@ -273,6 +273,17 @@ void increment(Collection<Field<?>> keys, String column, long amount)
...
+  /**
+   * Updates the specific fields in a range of rows from the table
+   *
+   * @param keyRange key range of the rows to update: cab only be a primary key prefix
+   * @param fields the fields to write to
+   * @throws InvalidFieldException if any of the keys are not part of table schema,
+   *                               or their types do not match the schema
+   * @throws IOException if there is an error reading from or updating the table
+   */
+  void updateAll(Range keyRange, Collection<Field<?>> fields) throws InvalidFieldException, IOException;
```
Comment (de-lan) [inline cdap-storage-spi/src/main/java/io/cdap/cdap/spi/data/StructuredTable.java:285]: nit, wondering if `updateRange` sounds more proper?

File: cdap-storage-spi/src/test/java/io/cdap/cdap/spi/data/StructuredTableTest.java [hunk 17]
```diff
@@ -1590,6 +1590,146 @@ public void testUpdate() throws Exception {
...
+                                                   Fields.bytesField(BYTES_COL, Bytes.toBytes("bytes-7")));
+    Assert.assertEquals(new HashSet<>(beforeUpdateSublist.get(0)), new HashSet<>(expectedFields1));
+    Assert.assertEquals(new HashSet<>(beforeUpdateSublist.get(1)), new HashSet<>(expectedFields2));
+
+    List<Field<?>> updates = Arrays.asList(Fields.stringField(STRING_COL, "val0"),
+                                           Fields.floatField(FLOAT_COL, (float) 0.0));
+    Range range = Range.create(Arrays.asList(Fields.intField(KEY, 6), Fields.longField(KEY2, 6L)),
+                               Range.Bound.INCLUSIVE,
+                               Arrays.asList(Fields.intField(KEY, 8), Fields.longField(KEY2, 8L)),
+                               Range.Bound.INCLUSIVE);
```
Comment (de-lan) [inline cdap-storage-spi/src/test/java/io/cdap/cdap/spi/data/StructuredTableTest.java]: I guess this should be `Range.Bound.EXCLUSIVE`

**Review by de-lan — APPROVED** [review]
some nit, otherwise LGTM

File: cdap-storage-spi/src/main/java/io/cdap/cdap/spi/data/StructuredTable.java [hunk 18]
```diff
@@ -273,6 +273,17 @@ void increment(Collection<Field<?>> keys, String column, long amount)
    */
   void deleteAll(Range keyRange) throws InvalidFieldException, IOException;
 
+  /**
+   * Updates the specific fields in a range of rows from the table
+   *
+   * @param keyRange key range of the rows to update: cab only be a primary key prefix
```
Comment (de-lan) [inline cdap-storage-spi/src/main/java/io/cdap/cdap/spi/data/StructuredTable.java:279]: nit: `cab` --> `can`

File: cdap-storage-spi/src/main/java/io/cdap/cdap/spi/data/StructuredTable.java [hunk 19]
```diff
@@ -273,6 +273,17 @@ void increment(Collection<Field<?>> keys, String column, long amount)
...
+   * @param fields the fields to write to
```
Comment (de-lan) [inline cdap-storage-spi/src/main/java/io/cdap/cdap/spi/data/StructuredTable.java:280]: nit: `the fields to write to` --> `the fields to update`

File: cdap-storage-spi/src/main/java/io/cdap/cdap/spi/data/StructuredTable.java (hunk 16 above)
Comment (sanjanasandeep) [reply cdap-storage-spi/src/main/java/io/cdap/cdap/spi/data/StructuredTable.java:285]: yup agree

Comment (sanjanasandeep) [reply cdap-storage-spi/src/main/java/io/cdap/cdap/spi/data/StructuredTable.java:285]: however we have deleteAll that also deletes range - so in terms of consistency it is better?

Comment (de-lan) [reply cdap-storage-spi/src/main/java/io/cdap/cdap/spi/data/StructuredTable.java:285]: yah it makes sense to be consistent

**Review by de-lan — APPROVED** [review]
